src/MagnumPlugins/ObjImporter/ObjImporter.cpp

```diff
@@ -115,387 +372,575 @@ ObjImporter::~ObjImporter() = default;
-        /* The previous object might end at the beginning of this line */
-        const std::streampos end = _file->in->tellg();
+        ensureMeshData();
+
+        ObjMesh* mesh;
+        if(sectionPrimitive == 'p') {
+            if(!meshData->points) {
+                // TODO: C++ 17
+                _state->meshes.emplace_back(new ObjMesh{*meshData, MeshPrimitive::Points});
+                meshData->points = _state->meshes.back().get();
+            }
+            mesh = meshData->points;
+        } else if(sectionPrimitive == 'l') {
+            if(!meshData->lines) {
+                // TODO: C++ 17
+                _state->meshes.emplace_back(new ObjMesh{*meshData, MeshPrimitive::Lines});
+                meshData->lines = _state->meshes.back().get();
+            }
+            mesh = meshData->lines;
+        } else { /* sectionPrimitive == 'f' */
+            if(!meshData->faces) {
+                // TODO: C++ 17
+                _state->meshes.emplace_back(new ObjMesh{*meshData, MeshPrimitive::Triangles});
+                meshData->faces = _state->meshes.back().get();
+            }
+            mesh = meshData->faces;
+        }
+        if(line.data() == nullptr) {
+            /* Usually for the last line */
+            mesh->sections.push_back(section);
+        } else {
+            mesh->sections.emplace_back(section.data(), size_t(line.data()-section.data()));
+        }
+        mesh->minPrimitives += minSectionPrimitives;
+
+        section = {nullptr};
+        minSectionPrimitives = 0;
+    };
+
+    auto finishObject = [&]{
+        if(object && object->groups.empty()) {
+            /* Create dummy mesh for this object so that it gets loaded as ObjectData */
+            _state->meshlessObjects.push_back(object->name);
+        }
+    };
+
+    while(!pos.empty()) {
 
         /* Comment line */
-        if(_file->in->peek() == '#') {
-            ignoreLine(*_file->in);
+        if(pos[0] == '#') {
+            finishSection();
+            line = pos = skipWhitespaces(ignoreLine(pos));
             continue;
         }
 
         /* Parse the keyword */
         std::string keyword;
-        *_file->in >> keyword;
+        ArrayView<char> word;
+        std::tie(word, pos) = nextWord(pos);
+        keyword = std::string(word.data(), word.size());
 
-        /* Mesh name */
-        if(keyword == "o") {
-            std::string name;
-            std::getline(*_file->in, name);
-            name = Utility::String::trim(name);
+        pos = skipWhitespaces(pos);
+        bool sectionEnd = true;
 
-            /* This is the name of first mesh */
-            if(thisIsFirstMeshAndItHasNoData) {
-                thisIsFirstMeshAndItHasNoData = false;
+        /* Vertex position */
+        if(keyword == "v") {
+            // TODO C++17: Could be one slick line with emplace_back ref return.
+            Vector3 v;
+            pos = getVector<3>(pos, v);
+            _state->positions.push_back(v);
```

Comment:
What about ~~`getVector`~~ `parseVector()` returning `Vector3` and modifying `pos` via a reference? Then it could fit on one line ;)